lib/lsp-server/src/lib.rs

```diff
@@ -183,67 +263,153 @@ impl Connection {
     ///    Ok(())
     /// }
     /// ```
-    pub fn initialize(
+    fn base_initialize<StartF, FinishF>(
         &self,
         server_capabilities: serde_json::Value,
-    ) -> Result<serde_json::Value, ProtocolError> {
-        let (id, params) = self.initialize_start()?;
+        start: StartF,
+        finish: FinishF,
+    ) -> Result<serde_json::Value, ProtocolError>
+    where
+        StartF: FnOnce(&Self) -> Result<(RequestId, serde_json::Value), ProtocolError>,
+        FinishF: FnOnce(&Self, RequestId, serde_json::Value) -> Result<(), ProtocolError>,
+    {
+        let (id, params) = start(self)?;
 
         let initialize_data = serde_json::json!({
             "capabilities": server_capabilities,
         });
 
-        self.initialize_finish(id, initialize_data)?;
+        finish(self, id, initialize_data)?;
 
         Ok(params)
     }
 
+    pub fn initialize(
+        &self,
+        server_capabilities: serde_json::Value,
+    ) -> Result<serde_json::Value, ProtocolError> {
+        self.base_initialize(
+            server_capabilities,
+            Connection::initialize_start,
+            Connection::initialize_finish,
+        )
+    }
```

Comment:
FWIW, I've long been thingking that initialize flow in the lsp-server is harmful. 

Like, yest, it's the same across all of the servers, so it _could_ be abtracted, but also, like, it's not a hard thing to do yourself, and, if you have initialize in your analogue of the mainloop, then it becomes much easier to debug this!

So I'd consider just axing any builtin support for initialization from the library